asserts/asserts.go

```diff
@@ -37,21 +43,27 @@ type AssertionType struct {
 	PrimaryKey []string
 
 	assembler func(assert assertionBase) (Assertion, error)
+	flags     typeFlags
 }
 
 // Understood assertion types.
 var (
-	AccountType         = &AssertionType{"account", []string{"account-id"}, assembleAccount}
-	AccountKeyType      = &AssertionType{"account-key", []string{"account-id", "public-key-id"}, assembleAccountKey}
-	ModelType           = &AssertionType{"model", []string{"series", "brand-id", "model"}, assembleModel}
-	SerialType          = &AssertionType{"serial", []string{"brand-id", "model", "serial"}, assembleSerial}
-	SnapDeclarationType = &AssertionType{"snap-declaration", []string{"series", "snap-id"}, assembleSnapDeclaration}
-	SnapBuildType       = &AssertionType{"snap-build", []string{"snap-sha3-384"}, assembleSnapBuild}
-	SnapRevisionType    = &AssertionType{"snap-revision", []string{"snap-sha3-384"}, assembleSnapRevision}
+	AccountType         = &AssertionType{"account", []string{"account-id"}, assembleAccount, 0}
+	AccountKeyType      = &AssertionType{"account-key", []string{"account-id", "public-key-id"}, assembleAccountKey, 0}
+	ModelType           = &AssertionType{"model", []string{"series", "brand-id", "model"}, assembleModel, 0}
+	SerialType          = &AssertionType{"serial", []string{"brand-id", "model", "serial"}, assembleSerial, 0}
+	SnapDeclarationType = &AssertionType{"snap-declaration", []string{"series", "snap-id"}, assembleSnapDeclaration, 0}
+	SnapBuildType       = &AssertionType{"snap-build", []string{"snap-sha3-384"}, assembleSnapBuild, 0}
+	SnapRevisionType    = &AssertionType{"snap-revision", []string{"snap-sha3-384"}, assembleSnapRevision, 0}
 
 // ...
 )
 
+// Freestanding assertion types (on the wire and/or self-signed).
+var (
+	SerialRequestType = &AssertionType{"serial-request", nil, assembleSerialRequest, freestanding}
```

Comment:
Not sure about this but ...

Just wondering if we need the `freestanding` flag when having no public key means it cannot be stored and is therefore freestanding?

The flag is more explicit but it would also be nice to have exactly one way to switch on these types of assertions. Most tests use the flag but Database.Add tests the primary key's length. Even the name of a new test, TestAddFreestandingNoPrimaryKey, suggests they're (currently) equivalent.